### PR TITLE
Add new updated limit method that takes word boundries into consideration

### PR DESCRIPTION
**Pull request by itsmewes**
Currently for limiting string lengths you can use `Str::limit` and `Str::words` which are great options. The problem that I ran into is that I would like to limit the string to whole words. `Str::limit` can cut into a word and `Str::words` can be tough to use since the length of a string  is wildly different depending on word sizes. This new `Str::limitTillWord` takes in characters like `Str::limit` but will make sure that it doesn't cut into any new words but completes on a word boundary. I believe this makes string lengths a lot more consistent when working with word boundaries. The downside to this is that there is no `Multibyte String (mb_*)` string version of `wordwrap`.

### CODE REVIEW

**Comment by ahinkle** [comment]
Seems a little niche. Might be better in userland. 👍

**Comment by henzeb** [comment]
Sounds interesting. The name is a bit lengthy ...

**Comment by itsmewes** [comment]
> Seems a little niche. Might be better in userland. 👍

That's a fair point. Personally I prefer having the limit always be on a word boundry but that is personal preference.

**Comment by itsmewes** [comment]
> Sounds interesting. The name is a bit lengthy ...

Yeah, naming is hard. Open to new names 🙂

**Comment by antonkomarev** [comment]
May it be an extra enum param in limit method? Default to limit as it works now, another value to trim to words. But maybe it's better to change signature in 12.x

**Comment by rodrigopedra** [comment]
`Str::wordLimit()`?

Really like it, I use a much worse implementation in many projects 😅 (thus the naming suggestion)

So, even if this doesn't make the cut, thank you so much for sharing

Specially for sending error reports through e-mail, or logging errors into constrained outputs, limiting by words is more helpful

File: src/Illuminate/Support/Str.php [hunk 1]
```diff
@@ -636,6 +636,27 @@ public static function limit($value, $limit = 100, $end = '...')
...
+    {
+        if (mb_strwidth($value, 'UTF-8') <= $limit) {
+            return $value;
+        }
+
+        $str = trim(preg_replace('/[\n\r]+/', ' ', strip_tags($value)));
+        $wordWrap = wordwrap($str, $limit);
+        $sentences = explode("\n", $wordWrap);
+
+        return preg_replace('/\W$/', '', $sentences[0]) . $end;
```
Comment (itsmewes) [inline src/Illuminate/Support/Str.php:657]: This is a proposed alternative to this as well which I believe would handle multibyte strings.

```suggestion
        if (mb_strwidth($value, 'UTF-8') <= $limit) {
            return $value;
        }

        $value = trim(preg_replace('/[\n\r]+/', ' ', strip_tags($value)));
        $trimmed = rtrim(mb_strimwidth($value, 0, $limit, '', 'UTF-8'));

        if (mb_substr($value, $limit, 1, 'UTF-8') === ' ') {
            return $trimmed.$end;
        }

        return preg_replace("/(.*)\s.*/", '$1', $trimmed).$end;
```

**Comment by itsmewes** [comment]
> May it be an extra enum param in limit method? Default to limit as it works now, another value to trim to words. But maybe it's better to change signature in 12.x
> 
> Something like `$mode`

This was actually my first thought, but then decided to not touch what was already there. I do think having some form of flag on the limit method would be first prize.

**Comment by crynobone** [comment]
Hi @itsmewes, can you rebase this PR to latest upstream

**Comment by taylorotwell** [comment]
What about multi-byte strings? I would almost just make this part of `Str::limit` and rely on named arguments to enable it...

`Str::limit($value, 100, completeWords: true)`

**Comment by itsmewes** [comment]
> What about multi-byte strings? I would almost just make this part of `Str::limit` and rely on named arguments to enable it...
> 
> `Str::limit($value, 100, completeWords: true)`

I have a comment that could work with multi-byte strings. I think having this as a named argument in `Str::limit` would be the best use case. I'll update the PR to take these into account. Thanks for the response!

**Comment by itsmewes** [comment]
> What about multi-byte strings? I would almost just make this part of `Str::limit` and rely on named arguments to enable it...
> 
> `Str::limit($value, 100, completeWords: true)`

This has been updated 😊

**Comment by driesvints** [comment]
Something went wrong with branching here it seems.

**Pull request closed without merging**